### PR TITLE
Invoker modularization

### PR DESCRIPTION
**Pull request by JosepSampe**
Moved the invoker as a module, so that it can be easy to implement and integrate other invokers.

Other improvements:
- Fixed minor issue with invocations
- Fixed unneeded memory usage
- store runtime_metadata into a local cache in order to reduce exec time
- Allowed to set the memory size directly in the call_async(), map() and map_reduce() methods. This makes more fine control of the memory used. 

It is now possible to use different memories with same executor, for example:
```
pw = pywren.ibm_cf_executor()
pw.map(my_function, range(total), runtime_memory=256)
pw.map(my_function, range(total), runtime_memory=512)
pw.map(my_function, range(total), runtime_memory=128)
```
Also, it is now possible to set different memory sizes in map_reduce method for the map and reduce phases, for example:
```
pw = pywren.ibm_cf_executor()
pw.map_reduce(my_function, range(total), my_reduce_function, map_runtime_memory=512, reduce_runtime_memory=128)
```

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by gilv** [comment]
@JosepSampe just noticed this one...only 26 files changed ;) is this something ready to test or still work in progress?

**Comment by JosepSampe** [comment]
Now it is ready to test. 
All tests passed in my case, but need you to verify it with your use cases ;)

File: pywren_ibm_cloud/libs/ibm/cloudfunctions_client.py [hunk 1]
```diff
@@ -192,56 +190,65 @@ def create_package(self, package):
...
-            conn.close()
-            logger.debug(str(e))
-            return self.request_invoke(action_name, payload)
+            if not self.is_cf_cluster:
+                conn.close()
+            log_msg = ('ExecutorID {} - Function {} invocation failed: {}'.format(exec_id, call_id, str(e)))
+            logger.debug(log_msg)
+            if self_invoked:
+                return None
+            return self.invoke(action_name, payload, self_invoked=True)
```
Comment (LachlanStuart) [inline pywren_ibm_cloud/libs/ibm/cloudfunctions_client.py]: Missing `package` first parameter on this line

Comment (gilv) [reply pywren_ibm_cloud/libs/ibm/cloudfunctions_client.py]: @JosepSampe ? ;)

Comment (JosepSampe) [reply pywren_ibm_cloud/libs/ibm/cloudfunctions_client.py]: Fixed

**Comment by gilv** [comment]
@JosepSampe @LachlanStuart  I merge this into master..master actually is 1.0.16-snapshot. So assuming Python doesn't mess packages, we can test things with release 1.0.15 and master 1.0.16-snapshot

**Pull request merged**